File: core/play/src/main/java/play/mvc/Security.java
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package play.mvc;

import play.inject.Injector;
import play.libs.typedmap.TypedKey;
import play.mvc.Http.Context;
import play.mvc.Http.Request;

import javax.inject.Inject;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;

/**
 * Defines several security helpers.
 */
public class Security {

    public static final TypedKey<String> USERNAME = TypedKey.create("username");

    /**
     * Wraps the annotated action in an {@link AuthenticatedAction}.
     */
    @With(AuthenticatedAction.class)
    @Target({ElementType.TYPE, ElementType.METHOD})
    @Retention(RetentionPolicy.RUNTIME)
    public @interface Authenticated {
        Class<? extends Authenticator> value() default Authenticator.class;
    }

    /**
     * Wraps another action, allowing only authenticated HTTP requests.
     * <p>
     * The user name is retrieved from the session cookie, and added to the HTTP request's
     * <code>username</code> attribute.
     */
    public static class AuthenticatedAction extends Action<Authenticated> {

        private final Function<Authenticated, Authenticator> configurator;

        @Inject
        public AuthenticatedAction(Injector injector) {
            this(authenticated -> injector.instanceOf(authenticated.value()));
        }

        public AuthenticatedAction(Authenticator authenticator) {
            this(authenticated -> authenticator);
        }

        public AuthenticatedAction(Function<Authenticated, Authenticator> configurator) {
            this.configurator = configurator;
        }

        public CompletionStage<Result> call(final Request req) {
            Authenticator authenticator = configurator.apply(configuration);
            return authenticator.getUsername(req)
                .map(username -> delegate.call(req.addAttr(USERNAME, username)))
                .orElseGet(() -> CompletableFuture.completedFuture(authenticator.onUnauthorized(req)));
        }

    }

    /**
     * Handles authentication.
     */
    public static class Authenticator extends Results {

        /**
         * Retrieves the username from the HTTP context; the default is to read from the session cookie.
         *
         * @param ctx the current request context
         * @return null if the user is not authenticated.
         *
         * @deprecated Deprecated as of 2.7.0. Use {@link #getUsername(Request)} instead.
         */
        @Deprecated
        public String getUsername(Context ctx) {
            return ctx.session().get("username");
        }

        /**
         * Retrieves the username from the HTTP request; the default is to read from the session cookie.
         *
         * @param req the current request
         * @return the username if the user is authenticated.
         */
        public Optional<String> getUsername(Request req) {
            return req.session().getOptional("username");
        }

        /**
         * Generates an alternative result if the user is not authenticated; the default a simple '401 Not Authorized' page.
         *
         * @param ctx the current request context
         * @return a <code>401 Not Authorized</code> result
         *
         * @deprecated Deprecated as of 2.7.0. Use {@link #onUnauthorized(Request)} instead.
         */
        @Deprecated
        public Result onUnauthorized(Context ctx) {
            return onUnauthorized(ctx.request());
        }

        /**
         * Generates an alternative result if the user is not authenticated; the default a simple '401 Not Authorized' page.
         *
         * @param req the current request
         * @return a <code>401 Not Authorized</code> result
         */
        public Result onUnauthorized(Request req) {
            return unauthorized(views.html.defaultpages.unauthorized.render(req.asScala()));
        }

    }


}

File: core/play/src/main/scala/play/api/mvc/Security.scala
/*
 * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.mvc

import play.api._
import play.api.libs.streams.Accumulator
import play.api.mvc.Results._

import scala.concurrent.ExecutionContext
import scala.concurrent.Future
import scala.language.reflectiveCalls
import scala.util.Failure
import scala.util.Success

/**
 * Helpers to create secure actions.
 */
object Security {
  private val logger = Logger(getClass)

  /**
   * The default error response for an unauthorized request; used multiple places here
   */
  private val DefaultUnauthorized: RequestHeader => Result = implicit request =>
    Unauthorized(views.html.defaultpages.unauthorized())

  /**
   * Wraps another action, allowing only authenticated HTTP requests.
   * Furthermore, it lets users to configure where to retrieve the user info from
   * and what to do in case unsuccessful authentication
   *
   * For example:
   * {{{
   *  //in a Security trait
   *  def username(request: RequestHeader) = request.session.get("email")
   *  def onUnauthorized(request: RequestHeader) = Results.Redirect(routes.Application.login)
   *  def isAuthenticated(f: => String => Request[AnyContent] => Result) = {
   *    Authenticated(username, onUnauthorized) { user =>
   *      Action(request => f(user)(request))
   *    }
   *  }
   * //then in a controller
   * def index = isAuthenticated { username => implicit request =>
   *     Ok("Hello " + username)
   * }
   * }}}
   *
   * @tparam A the type of the user info value (e.g. `String` if user info consists only in a user name)
   * @param userinfo function used to retrieve the user info from the request header
   * @param onUnauthorized function used to generate alternative result if the user is not authenticated
   * @param action the action to wrap
   */
  def Authenticated[A](
      userinfo: RequestHeader => Option[A],
      onUnauthorized: RequestHeader => Result
  )(action: A => EssentialAction): EssentialAction = {
    EssentialAction { request =>
      userinfo(request)
        .map { user =>
          action(user)(request)
        }
        .getOrElse {
          Accumulator.done(onUnauthorized(request))
        }
    }
  }

  def WithAuthentication[A](
      userinfo: RequestHeader => Option[A]
  )(action: A => EssentialAction): EssentialAction = {
    Authenticated(userinfo, DefaultUnauthorized)(action)
  }

  /**
   * An authenticated request
   *
   * @param user The user that made the request
   */
  class AuthenticatedRequest[+A, U](val user: U, request: Request[A]) extends WrappedRequest[A](request) {
    protected override def newWrapper[B](newRequest: Request[B]): AuthenticatedRequest[B, U] =
      new AuthenticatedRequest[B, U](user, newRequest)
  }

  /**
   * An authenticated action builder.
   *
   * This can be used to create an action builder, like so:
   *
   * {{{
   * class UserAuthenticatedBuilder (parser: BodyParser[AnyContent])(implicit ec: ExecutionContext)
   *   extends AuthenticatedBuilder[User]({ req: RequestHeader =>
   *   req.session.get("user").map(User)
   * }, parser) {
   *   @Inject()
   *   def this(parser: BodyParsers.Default)(implicit ec: ExecutionContext) = {
   *     this(parser: BodyParser[AnyContent])
   *   }
   * }
   * }}}
   *
   * You can then use the authenticated builder with other action builders, i.e. to use a
   * messagesApi with authentication, you can add:
   *
   * {{{
   *  class AuthMessagesRequest[A](val user: User,
   *                              messagesApi: MessagesApi,
   *                              request: Request[A])
   * extends MessagesRequest[A](request, messagesApi)
   *
   * class AuthenticatedActionBuilder(val parser: BodyParser[AnyContent],
   *                                  messagesApi: MessagesApi,
   *                                  builder: AuthenticatedBuilder[User])
   *                                 (implicit val executionContext: ExecutionContext)
   *     extends ActionBuilder[AuthMessagesRequest, AnyContent] {
   *   type ResultBlock[A] = (AuthMessagesRequest[A]) => Future[Result]
   *
   *   @Inject
   *   def this(parser: BodyParsers.Default,
   *            messagesApi: MessagesApi,
   *            builder: UserAuthenticatedBuilder)(implicit ec: ExecutionContext) = {
   *     this(parser: BodyParser[AnyContent], messagesApi, builder)
   *   }
   *
   *   def invokeBlock[A](request: Request[A], block: ResultBlock[A]): Future[Result] = {
   *     builder.authenticate(request, { authRequest: AuthenticatedRequest[A, User] =>
   *       block(new AuthMessagesRequest[A](authRequest.user, messagesApi, request))
   *     })
   *   }
   * }
   * }}}
   *
   * @param userinfo The function that looks up the user info.
   * @param onUnauthorized The function to get the result for when no authenticated user can be found.
   */
  class AuthenticatedBuilder[U](
      userinfo: RequestHeader => Option[U],
      defaultParser: BodyParser[AnyContent],
      onUnauthorized: RequestHeader => Result = implicit request => Unauthorized(views.html.defaultpages.unauthorized())
  )(implicit val executionContext: ExecutionContext)
      extends ActionBuilder[({ type R[A] = AuthenticatedRequest[A, U] })#R, AnyContent] {
    lazy val parser = defaultParser

    def invokeBlock[A](request: Request[A], block: (AuthenticatedRequest[A, U]) => Future[Result]) =
      authenticate(request, block)

    /**
     * Authenticate the given block.
     */
    def authenticate[A](request: Request[A], block: (AuthenticatedRequest[A, U]) => Future[Result]) = {
      userinfo(request)
        .map { user =>
          block(new AuthenticatedRequest(user, request))
        }
        .getOrElse {
          Future.successful(onUnauthorized(request))
        }
    }
  }

  object AuthenticatedBuilder {

    /**
     * Create an authenticated builder
     *
     * @param userinfo The function that looks up the user info.
     * @param onUnauthorized The function to get the result for when no authenticated user can be found.
     */
    def apply[U](
        userinfo: RequestHeader => Option[U],
        defaultParser: BodyParser[AnyContent],
        onUnauthorized: RequestHeader => Result = DefaultUnauthorized
    )(implicit ec: ExecutionContext): AuthenticatedBuilder[U] = {
      new AuthenticatedBuilder(userinfo, defaultParser, onUnauthorized)
    }
  }
}
